building: add Linux userspace examples

johndoe31415 opened the pull request:
Adds a typical use-case scenario where OP-TEE's PKCS#11 interface is used to create a TLS connection using client-certificate based authentication. Keys are stored inside OP-TEE.

johndoe31415 commented:
Also another note, I think it's justified to add a "integration" chapter because I have in mind to also add some details regarding key sealing/key wrapping and its integration into dm-crypt. But I've not yet had time to write that up.

jforissier commented on building/userland_integration.rst:
```diff
@@ -0,0 +1,388 @@
+    Library          OP-TEE PKCS11 Cryptoki library (ver 0.1)
+    Using slot 0 with a present token (0x0)
+
+
```
Remove one blank line

jforissier reviewed (commented):
Hi @johndoe31415,

Thanks for this doc update. FYI, I have been working on very similar stuff recently to enable PKCS#11 with AWS Greengrass. Aside from Greengrass-specific configuration, the only difference with what you documented here is that the certificate is signed by the Amazon CA, then it needs to be imported into PKCS#11 (`pkcs11-tool ... -write-object device.crt`).

Anyways this looks good, please see a few comments below.

johndoe31415 commented:
Hey @jforissier, thanks for the super quick feedback! You're making great points, I'll include them tonight when I have some time and squash/forcepush. Happy to contibute something back after the great help that many (including you) have provided to me in the past.

johndoe31415 commented:
@jforissier All your comments should have been addressed/clarified. I also fixed two typos and changed the syntax highlighting from "bash" to "none" (since otherwise root-terminal lines starting with "#" would be interpreted as comments and OpenSSL output "return" would be highlighted as a keyword). I'm leaving this as individual commits so you can re-review. Once you're happy, I'll squash them. Have a good evening!

jforissier commented on building/userland_integration.rst:
```diff
@@ -30,29 +30,41 @@ part of the `OP-TEE client tools <https://github.com/OP-TEE/optee_client>`_.
 
+.. hint::
+
+   When testing OP-TEE under QEmu, OpenSC should be built by default as well
```
QEMU (all uppercase)

jforissier commented on building/userland_integration.rst:
```diff
@@ -30,29 +30,41 @@ part of the `OP-TEE client tools <https://github.com/OP-TEE/optee_client>`_.
-    # CKTEEKSO=/usr/local/lib/libckteec.so
-    # pkcs11-tool --module ${CKTEEKSO} --show-info
+    # alias p11="pkcs11-tool --module /usr/local/lib/libckteec.so"
+    # ${CKTEEKSO} --show-info
```
`# p11 --show-info`

jforissier commented on building/userland_integration.rst:
```diff
@@ -218,16 +232,30 @@ For this, we create a small configuration file for OpenSSL and call it ``optee_h
 
+.. hint::
+
+   When testing OP-TEE under QEmu, libp11 is not compiled by default. For easy
```
QEMU

jforissier commented on building/userland_integration.rst:
```diff
@@ -218,16 +232,30 @@ For this, we create a small configuration file for OpenSSL and call it ``optee_h
+       make BR2_PACKAGE_OPENSC=y BR2_PACKAGE_LIBOPENSSL=y BR2_PACKAGE_LIBOPENSSL_BIN=y BR2_PACKAGE_LIBP11=y
+
+   This will ensure that OpenSC (for the command line utility ``pkcs11-tool``),
+   OpenSSL, and libp11 are all built and installed in the QEmu environment.
```
QEMU

jforissier commented on building/userland_integration.rst:
```diff
@@ -30,29 +30,41 @@ part of the `OP-TEE client tools <https://github.com/OP-TEE/optee_client>`_.
+``pkcs11-tool`` of the `OpenSC project <https://github.com/OpenSC/OpenSC>`_ to
+initiate first communication with the emulated smart card.  In the following,
+we assume that ``libckteec.so`` has been installed in
+``/usr/local/lib/libckteec.so``. For simplicity reasons, we define an alias to
```
Actually in the QEMU environment (at least QEMUv8), it is `/usr/lib/libckteec.so.0`. Should this path be used instead?

jforissier reviewed (commented):
Thanks for the update. A few more things...

johndoe31415 commented:
Should be all addressed. Yes, `/usr/local/lib` was correct in my case since I built optee_client myself. But `/usr/lib` is probably better.

jforissier commented:
> Should be all addressed. 

Very good. That's a helpful document I wish I could have found a week ago or so :wink: 

> Yes, `/usr/local/lib` was correct in my case since I built optee_client myself. But `/usr/lib` is probably better.

On a related note, I proposed `/usr/lib/libckteec.so.0` (with the `.0` suffix) because for some reason the `.so` is not installed in the Yocto-based build that I have used for my own tests. But since it is present in the QEMU builds and is usually how PKCS#11 modules are named, I believe the path you have now used in the latest version (`/usr/lib/libckteec.so`) makes the most sense.

Please squash the fixup commits and apply:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

johndoe31415 commented:
> > Should be all addressed.
> 
> Very good. That's a helpful document I wish I could have found a week ago or so wink

Thanks!

> > Yes, `/usr/local/lib` was correct in my case since I built optee_client myself. But `/usr/lib` is probably better.
> 
> On a related note, I proposed `/usr/lib/libckteec.so.0` (with the `.0` suffix) because for some reason the `.so` is not installed in the Yocto-based build that I have used for my own tests. But since it is present in the QEMU builds and is usually how PKCS#11 modules are named, I believe the path you have now used in the latest version (`/usr/lib/libckteec.so`) makes the most sense.

Fair enough. Yep I think the entry of barrier is probably the lowest when it just works out-of-the-box in the QEMU build.

> Please squash the fixup commits and apply:

Done!

etienne-lms commented on building/userland_integration.rst:
```diff
@@ -0,0 +1,416 @@
+
+Note that the public key matches exactly that which we have previously created
+(``04 e3 f8...``). This CSR could then be signed by a CA. For simplicity
+purposes, we can also use a self-signed certificate and sign with out own
```
s/out/our/

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` (a typo to fix)
Thanks for the detailed description.

johndoe31415 commented:
Thanks for the review, @etienne-lms. I've incorporated and made a fixup commit, adding your Ack.

One more thing, in the `optee_hsm.conf` I think it would be good to change `/usr/lib/arm-linux-gnueabihf/engines-1.1/libpkcs11.so` to `/usr/lib/engines-1.1/libpkcs11.so` (that's where it resides in the QEMU environment). Since both of you @etienne-lms @jforissier reviewed already I wanted to ask if that's okay with you first.

johndoe31415 commented:
Alright, made all changes and squashed the commit. Thanks for the quick review @jforissier and @etienne-lms!

etienne-lms reviewed (approved):
@jbech-linaro, @jenswi-linaro, @jforissier, Ok to merge this change?

johndoe31415 commented:
Ping on this PR... can we resurrect it? Thanks! :-)

jbech-linaro commented:
Hi @johndoe31415 , @etienne-lms , yes this went below my radar.  I see a readthedocs error, but that looks unrelated to this PR. I'll merge this one since it's been reviewed quite some time again. Thanks @johndoe31415 !

etienne-lms commented:
thanks @jbech-linaro
